### DISCUSSION TITLE
Project roadmap

### DISCUSSION THREAD

**Discussion by MarcinOrlowski** [logisim-evolution/logisim-evolution#695] [General]
As the development progress and with multiple changes (and planned changes like ie. #690 that can break backward compatibility) I believe it is high time to talk the project roadmap and long term goals. I got a feeling this is not precisely specified at the moment, so it would be great to try to have something for say at least next 6 months, as (IMHO :) uncoordinated development is counter-beneficial (so are "rolling deadlines") and all this leads to putting efforts into irrelevant actions while what's more important are left unattended or not addressed, breaking backward compatibility etc, not to mention scaring off potential contributors.

I call all persons who I see merged any code changes in last few months:
@BFH-ktt1 
@R3dst0ne
@maehne
@davidhutchens 
@zdimension 
@GavinAndrews
@simoc

**Comment by BFH-ktt1** [comment]
We should define a time-slot for this when all are in a situation to have some slack to discuss and think. Things like PR-roles and methods should also be included. I would propose that each of us defines when would be a good time, and than afterwards maybe by which "medium" we want to discuss

**Reply by maehne** [reply]
I agree, we should postpone this discussion after the exam (and probably vacation) period. Starting from calendar week 32, I will be more available. In the meantime, we can collect in this thread ideas for goals we would like to work on in the short and mid term.

**Comment by BFH-ktt1** [comment]
Just to put some discussion material, here a list (probably I still forgot some points) with basically my action points:
Aditions:
1) As we developed a new HW platform (leguan) we need for our courses till mid-september an interface from LED/RGB-LED/LED-Array to an RGB-LED-Array that is controlled row-column wise in scanning mode. This implies a HDL-generator part that is invisible for the user and transforms for example 1 LED to a given LED in the array. It requires a new concept in the HDL-generator/mapping part.
2) As part of the merging effort to merge back the great additions of @kevinawalsh , there is one very nice part still open, n.l. copy-paste over different instances of logisim (this is currently not yet possible). My progress on this merging effort I documented in #231.
3) Adding FSM-support for logisim, meaning drawing state-diagrams and than automatically generate the circuit for it using a user-selectable coding table. Also the inverse would be nice, taking a circuit of an FSM and draw the state-diagram from it. There has already been a very nice effort in this direction @sderrien but we never finished it.

bug-fixes:
1) The splitter is the most difficult component in logisim concerning netlist-generation, and there is still a bug in this part (see #569)
2) The HDL-generator for the LED-Array works in most cases, but also still contains some bugs for the corner cases (one-line, one column).
3) The chronogram was already ported from the holycross version, but @kevinawalsh pushed some bug-fixes to the chronogram that were not yet merged.

General:
1) Some of the documentation is very outdated (for example the one on the timing-diagramm, and simulator where it is stated that you need a clock-component to make it work, which is not anymore the case). To be able to continue adding features, I think we should first make sure that at least for one language (I propose English) we synchronize the documentation.
2) There are still a lot of issues open that should be addressed in the way that we should check if it is still an issue. And put a date for fixing them.

Most probably I forgot a lot of points, I just wanted to clear my head here before I forget

**Reply by sderrien** [reply]
Hello,

I would be glad to help integrating the FSM part, as it would clearly benefit from a better integration. As far as I remember I
was stuck because of the gradle building system which I am not familiar with, and which was difficult to adapt to my code base which uses the Xtend/Xtext framework from EMF.  Integrating an FSM synthesis step would also be a big plus (I guess we could take advantage of the Karnaugh map classes already available).

**Reply by maehne** [reply]
@sderrien: Thanks a lot for offering your help to integrate your FSM part into Logisim-evolution! Indeed, figuring out how to use the Xtend/Xtext framework together with our `gradle` build system is one of the key issues. Given the passed time and amount of changes, which have been incorporated into Logisim-evolution over the past two years, it would be probably a lot of effort trying to merge commit by commit back into Logisim-evolution. Instead, taking a snapshot of your additions and integrating them in one step on top of our `develop` branch might be the simpler approach. Could you compile a list of all the source files, which you have touched in your fork to implement the FSM part? Are there other relevant improvements, which could be merged into Logisim-evolution?

**Reply by sderrien** [reply]
Sure, but we first need to be able to use gradle to build a project with multiple source folders, which was where we were stuck if I remember correctly. For that I would need some help, as I have no experience at all with gradle.

**Reply by maehne** [reply]
@Caellian: Could you provide maybe some support in this matter?

**Reply by Caellian** [reply]
## Gradle

Adding submodules shouldn't be hard (2 days of easy work tops) but the resulting jar file would still include everything.

I think Logisim-evolution would benefit from allowing users to choose certain features they need instead of bundling everything. This might be a bit nitpicky though as the jar size of latest release is ~70MB (FreeCAD is ~430MiB (more features); Dia is ~20MiB (less features)).

I haven't used much of what Logisim-evolution offers throughout my education (Verilog, FPGA, VHDL) nor in my free time, so there's a good chance others wouldn't either.

Here are some solutions anyway:

### Feature variants

Gradle added [feature variants](https://docs.gradle.org/current/userguide/feature_variants.html) since I last helped out with Logisim-evolution. These should allow modeling dependencies in a scalable way where they could be replaced with stubs in case they're not needed. I never used them though so I'd have to look into it.

Combining these with [project properties](https://docs.gradle.org/current/userguide/build_environment.html#sec:project_properties) should allow disabling certain features if they're not wanted by users.

### Plugins

Maybe a better solution would be to add proper support for plugins. This would require some (maybe a lot of) restructuring to support dynamically calling functions loaded from external jar files. I'd have to properly dive into the sources to figure out exactly how much of "if not available do nothing, if jar is loaded do this" is needed.

The upside is that this doesn't require stubs nor re-downloading the entire jar file in cases where the user didn't download distribution with wanted feature set in the first place.

You can create a JSON file in the root of the repository specifying which plugins are available, separate them into their own repositories, so that you don't need to host a server and something like [Eclipse marketplace](https://marketplace.eclipse.org/).

### Old (unfinished) solution

I previously wanted to write a Gradle plugin that adds something like feature flags from C++ to Java, but this really went against how Gradle works, required parsing Java sources, dealing with ASTs and I didn't have time to do all that and properly maintain it so I decided against it. This would be the worst solution out of all I listed as the unused sources would still be bundled in resulting jar file (they just wouldn't be accessible from GUI). I planned on mixing it with submodules to exclude those sources, but the result would be extremely complicated code and build script so I dropped it.

Let me know if you'd like help with any of the first two, I'd gladly help out.

## CI artifact publishing

GitHub can automatically publish build artifacts on successful builds and version releases. There's [actions/upload-artifact](https://github.com/actions/upload-artifact) action that makes this very simple. This would allow everyone to get the built distributions of latest development commit without having to compile it themselves. It would also reduce the work maintainers have to do while publishing releases as they'd automatically include built distributions.

Take a look at my workflow for a C++ project for reference [here](https://github.com/Caellian/java-decompiler/blob/master/.github/workflows/build-linux.yml). Only the build part would be different, instead of publishing a binary - jar file(s) can be published.

## JavaFX

This isn't very important, but would make the GUI look more "sleek" and "professional".

JavaFX has better support for animations and transitions and they're much easier to add than with Swing.

- Supports Swing subcomponents so complicated parts of the GUI code don't need to be rewritten.
- Allows very simple theming with CSS.
- Dark theme. I recently heard someone say their eyes hurt after designing an entire CPU in Logisim.

**Reply by MarcinOrlowski** [reply]
> This would allow everyone to get the built distributions of latest development commit 

That's almost here -> #818

**Comment by MarcinOrlowski** [comment]
Guys, let me be the one who pushes this forward a bit. I understand, holidays, sun and things as a big competitive force attracting your attention right now, but what if we try to a) plan some on-line meeting for all who interested to participate, b) plan it far ahead so it will be still to coordinate the date that fits all interested participants. c) I believe that for the first on-line meeting it should be fine to allocate 1 hours (2 hours max). It's enough to meet each other and the discuss the project. d) We shall have agenda clearly defined prior the meeting.

So to sum up. Whoever would like to participate, please specify the date range that works for you AND date range that won't work for you (if any). Also say what time would work for you (please mention timezone).

**Reply by zdimension** [reply]
I'm available during pretty much anytime after Aug 4th, during the day (Paris timezone, so currently CEST / GMT+2)

**Reply by MarcinOrlowski** [reply]
I'm available after Aug 14. I am flexible about hours (CEST)

**Reply by Caellian** [reply]
Can someone record the meeting and link it here?

I would like to hear your concerns with regards to project structure and Gradle so I have general guidelines on how to help. Specifically which parts of the code base should be separated into submodules (e.g. com.cburch.gray, rendering code, component libraries, ...).

**Reply by R3dst0ne** [reply]
For the next one and a half month, I'm pretty much always available. (CEST)

> Can someone record the meeting and link it here?

I'm strongly against this. But we could write a protocol with all the information.

**Reply by MarcinOrlowski** [reply]
I second @R3dst0ne. We can post "meeting minutes" once it's done, so @Caellian (and anyone interested) pleaase post bullet points of what you think should be part of the meeting agenda.

**Reply by Caellian** [reply]
From skimming over discussions and issues:

- Gradle submodules (what code should be separated from the main project; which forks to turn into submodules)
  - Using git submodules to keep commit histories, PRs, issues, etc. separated
  - Separating assets such as component libraries into their own repositories and providing a mechanism for updating only those without re-downloading the entire jar or publishing new versions on package managers
- UI, UX
  - Use `SystemLookAndFeel` LaF and fall back on Numbus and then on Metal (although Nimbus seems to be present everywhere)
    - GTK+ LaF on Linux might look bad if GTK is configured to use a dark theme and light theme icons are used
  - Transitioning to JavaFX (not really that important, but would be nice)
    - JavaFX UI can contain Swing components so a complete rewrite of everything at once isn't necessary
  - Dark theme
  - Icons (to SVG or not to SVG)
    - Theming would be simpler by allowing changing of fill and stroke colors in SVG documents (which isn't that hard)
    - SVG would make icons very easy to modify
  - Media materials - providing SVG sources for icons and art ([example 1](https://mozilla.design/mozilla/logo-usage/), [example 2](https://www.rust-lang.org/policies/media-guide)) in case someone wants to create a YT thumbnail, blog post, etc. containing them
- Simulation engine
  - Allow switching between different simulation algorithms in settings
    - Something like how games can have different rendering backend implementations for OGL, D3D11 & D3D12
    - Would make it easier to gradually improve existing algorithm implementations, to test them against each other and to experiment with new ones
    - Kind of fixes #833
  - Support for bidirectional pins
  - Other improvements?

I'm not all that familiar with the actual code of this project (I hope to fix that this summer 😃), but these are just some ideas that you might find useful.

**Reply by BFH-ktt1** [reply]
I'm available except for the week of August the 16th-20th. For the time I am flexible (CEST)

**Reply by davidhutchens** [reply]
I am flexible as to date and time and I can read a summary if I miss it. However, I am in US Eastern Daylight Time (EDT, GMT-4) so I am 6 hours behind most of you.

**Reply by maehne** [reply]
I am currently on vacation. I'll be available again starting from 11 August 2021. To accomodate for people from the US, like @davidhutchens, we should schedule the meeting in the late afternoon, i.e., after 16:00 o'clock CEST. We can host the meeting either through MS Teams or Jitsi Meet. How about having the meeting on Thursday, 12 August, starting from 16:15 o'clock (CEST)? (I have already a meeting before that should end 16:00 o'clock)? Alternative convenient dates for me would be Monday, 30 August and Tuesday, 31 August, starting from 16:00 o'clock.

**Reply by zdimension** [reply]
August 12 is OK for me, not sure for 30 and 31

**Reply by R3dst0ne** [reply]
12, 30 and 31 August are OK for me.

**Reply by zdimension** [reply]
So, Aug 12 at 16:15 CEST it is then?

**Reply by MarcinOrlowski** [reply]
As mentioned, I am off till 14th so that won't work for me. Then I saw @BFH-ktt1 is off 16-20th, so I believe once you put others' slots all into matrix the safest date will be like last week of Aug. I suggest we try to aim for either Aug 21st (Sat) or subsequent Saturdays (Aug 28th or Sep 4th) and I'd propose 1h (maybe 1,5h max) time slot, for the first meeting to meet'n'greet, talk some project stuff and plan something for the next one.

**Reply by zdimension** [reply]
My bad, I forgot about that. I'm available Aug 21, 28, 30, 31 and Sep 4.

**Reply by maehne** [reply]
Thanks for your replies! I am not available on weekends due to family obligations. I discussed with @BFH-ktt1 possible time slots based on your feedback and we would propose to do the meeting on Monday, 2021-08-30. We can be available from 15:00 to 17:30 CEST.

@MarcinOrlowski, @R3dst0ne, @davidhutchens, @zdimension, @simoc, @sderrien, @Caellian: 

- Who could participate on that date?
- Is 15:00 CEST OK for you to start or would you prefer to start later, e.g. 15:30 or 16:00 CEST?
- We can host the meeting on either our school's Jitsi server or Microsoft Teams. Which option do you prefer? As you are all off-premise from our perspective and we don't have your e-mail contacts (which we usually use for meeting invitations), we should perform a small test of the video conference platform in advance (e.g., next week) to ensure that it is working as expected on the meeting day with the provided link. Who would volunteer for that?

**Reply by zdimension** [reply]
15:00 CEST on 2021-08-30 is OK for me; I have no preference between Jitsi and Teams. I'm available next week for a video conference test.

**Reply by davidhutchens** [reply]
15:00 CEST on 2021-08-30 is OK for me. Either Jitsi or Teams should work. I can participate in a test next week.

**Reply by MarcinOrlowski** [reply]
15:00 CEST on 2021-08-30 is OK for me. I do not know what Jitsi is though. Teams is OK perhaps. So is Zoom or Hangouts.

**Reply by R3dst0ne** [reply]
15:00 CEST on 2021-08-30 is OK for me. I'm not a fan of Teams, but I'm willing to use it, if it is the best option.

**Comment by davidhutchens** [comment]
Are we meeting at 15:00 CEST on 2021-08-30? Which platform. It sounded like Teams was the choice, but I haven't seen any updates. How will we get connection information?

**Reply by zdimension** [reply]
I had written down that time too, I assumed maybe @BFH-ktt1 or @maehne would post the link of the meeting tomorrow

**Reply by MarcinOrlowski** [reply]
Yep, I was about to ask the same. Any tech details (I understood some of you want to host it due to existing experience). If we still need to set it up, it's high time :)

**Reply by BFH-ktt1** [reply]
As far as I know it takes place and @maehne will send a team's link

**Reply by maehne** [reply]
Sorry, forgot about sending the invitation due to other tasks. Yes, ist will take place tomorrow, Monday 2021-08-30 at 15:00 CEST, and I will send a Teams invitation this evening (i.e., when I can access my computer) to those I have an e-mail address and post the link to join here for the others. Looking forward to the discuss with all of you!

**Reply by maehne** [reply]
@BFH-ktt1, @MarcinOrlowski, @davidhutchens, @zdimension, @mbaillif, @R3dst0ne, @Caellian, @sderrien:    

Dear all, I sent the invitation to your e-mails, which you use for your Git commits. Just in case, here's the [link](https://teams.microsoft.com/l/meetup-join/19%3ameeting_OTIyMWE5ZjMtMTU0ZS00NzUxLTkzNWQtOGI4OTk1MmFiZTRh%40thread.v2/0?context=%7b%22Tid%22%3a%22d6a1cf8c-768e-4187-a738-b6e50c4deb4a%22%2c%22Oid%22%3a%22351b18be-5e39-4790-bb12-0058defdaa0d%22%7d) to join the meeting on MS Teams. Best regards, Torsten Maehne

**Reply by maehne** [reply]
If someone should have trouble joining, you can post here. We'll monitor the thread.

**Reply by zdimension** [reply]
Having Teams-related problems joining, shouldn't take long to solve.

**Reply by R3dst0ne** [reply]
We didn't discuss how we are going to combine and then releas our notes. ^^'

**Reply by zdimension** [reply]
Here are mine:
```
plugin system
-> FSM plugin by steven derrien

better documentation hosting
-> allow updating doc separately from app

better document disassembly

better tracing of circuits

javadoc
```

**Comment by MarcinOrlowski** [comment]
I created separate topic #972 related to our meeting. Please post there to keep related data in one bucket.
